crates/engine/invalid-block-hooks/src/witness.rs

```diff
@@ -1,13 +1,185 @@
+        //
+        // Note: We grab *all* accounts in the cache here, as the `BundleState` prunes
+        // referenced accounts + storage slots.
+        let mut hashed_state = HashedPostState::from_bundle_state(&bundle_state.state);
+        for (address, account) in db.cache.accounts {
+            let hashed_address = keccak256(address);
+            hashed_state
+                .accounts
+                .insert(hashed_address, account.account.as_ref().map(|a| a.info.clone().into()));
+
+            let storage = hashed_state
+                .storages
+                .entry(hashed_address)
+                .or_insert_with(|| HashedStorage::new(account.status.was_destroyed()));
+
+            if let Some(account) = account.account {
+                state_preimages.insert(hashed_address, alloy_rlp::encode(address).into());
+
+                for (slot, value) in account.storage {
+                    let slot = B256::from(slot);
+                    let hashed_slot = keccak256(slot);
+                    storage.storage.insert(hashed_slot, value);
+
+                    state_preimages.insert(hashed_slot, alloy_rlp::encode(slot).into());
+                }
+            }
+        }
+
+        // Generate an execution witness for the aggregated state of accessed accounts.
+        // Destruct the cache database to retrieve the state provider.
+        let state_provider = db.database.into_inner();
+        let witness = state_provider.witness(HashedPostState::default(), hashed_state.clone())?;
+
+        // Write the witness to the output directory.
+        let mut file = File::options()
+            .write(true)
+            .create_new(true)
+            .open(self.output_directory.join(format!("{}_{}.json", block.number, block.hash())))?;
+        let response = ExecutionWitness { witness, state_preimages: Some(state_preimages) };
+        file.write_all(serde_json::to_string(&response)?.as_bytes())?;
+
+        if cfg!(debug_assertions) {
+            // The bundle state after re-execution should match the original one.
+            pretty_assertions::assert_eq!(
+                bundle_state,
+                output.state,
+                "Bundle state mismatch after re-execution"
+            );
+
+            // Calculate the state root and trie updates after re-execution. They should match
+            // the original ones.
+            let (state_root, trie_output) = state_provider.state_root_with_updates(hashed_state)?;
+            if let Some(trie_updates) = trie_updates {
+                pretty_assertions::assert_eq!(
+                    state_root,
+                    trie_updates.1,
+                    "State root mismatch after re-execution"
+                );
+                pretty_assertions::assert_eq!(
+                    &trie_output,
+                    trie_updates.0,
+                    "Trie updates mismatch after re-execution"
+                );
+            }
+        }
```

Comment:
now that i think about it, we should do these sanity checks even without debug assertions on to prevent a situation when there's a witness mismatch but it's due to post state mismatch after re-execution